Forwarder: add KMS permission to get log from Encrypted S3 Bucket

guitarrapc opened the pull request:
<!--- Please remember to review the [contribution guidelines](https://github.com/DataDog/datadog-serverless-functions/blob/master/CONTRIBUTING.md) if you have not yet done so._  --->

### What does this PR do?

<!--- A brief description of the change being made with this pull request. --->

Fix Forwarder function fail to invoke `s3:GetObject` with Encrypted S3 Bucket.
Now Forwarder support Encrypted S3 Bucket.

### Motivation

<!--- What inspired you to submit this pull request? --->

If Trigger S3 bucket is encrypted, Lambda Execution IAM Role require KMS priviledges.

> Object is encrypted by AWS KMS - https://aws.amazon.com/premiumsupport/knowledge-center/s3-troubleshoot-403

### Testing Guidelines

<!--- How did you test this pull request? --->
Create S3 Bucket with KMS Encyrpted, 

* Deploy original CF and forwarder cause error as `s3:GetObject` Access Denied.

![image](https://user-images.githubusercontent.com/3856350/91332123-a7bf4780-e806-11ea-9f62-7e95391fed5b.png)

* Deploy new CF and forwarder run without error.

![image](https://user-images.githubusercontent.com/3856350/91332148-b279dc80-e806-11ea-987e-38f7794d239b.png)

### Additional Notes

<!--- Anything else we should know when reviewing? --->

### Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [ ] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [x] This PR passes the integration tests (ask a Datadog member to run the tests)
- [x] This PR passes the unit tests 
- [x] This PR passes the installation tests (ask a Datadog member to run the tests)

guitarrapc committed:
feat: add KMS permission to get log from Encrypted S3 Bucket

guitarrapc commented:
nothing change with failed lint files, though.

> https://github.com/DataDog/datadog-serverless-functions/pull/337/checks?check_run_id=1032593121

no idea with why failed.

> https://github.com/DataDog/datadog-serverless-functions/pull/337/checks?check_run_id=1032593147

tianchu commented:
@guitarrapc Thanks for your PR! Do you mind sharing the steps to reproduce the issue and a link to the article that you referenced? These help us perform a quick test on our side and evaluate the PR with confidence.

BTW, The integration tests were a little bit flaky and should have been fixed.

guitarrapc commented:
Thanks for the review.

## step to reproduce

In my case was CloudTrail with enctypted S3 Bucket.

1. Create S3 Bucket with KMS encypted.
1. Assign S3 Bucket with CloudTrail.
1. Create datadog serverless function.
1. Set CloudTrail S3 Bucket as trigger.
1. s3:GetObject error happen for default CloudFormation template.
1. Add KMS permission tp CloudFormation template and S3 Object successfully loaded by function.

Do you need terraform template or any reproduceable IoC?

tianchu commented on aws/logs_monitoring/template.yaml:
```diff
@@ -433,6 +433,14 @@ Resources:
+            - Effect: Allow
+              Action:
+                - kms:Encrypt
+                - kms:Decrypt
```
@guitarrapc According to my tests, if you use the AWS managed CMK (i.e., key `aws/s3`), no extra permissions are required for the Forwarder to download the object. If you use a customer managed CMK, the Forwarder only needs `kms:Decrypt` for downloading. I guess the other kms permissions you listed are required for uploading objects to the bucket, which means they are not needed by the Forwarder? Can you confirm and update the PR if only adding `kms:Decrypt` works for you as well? Thanks!

guitarrapc replied:
That make sence. AS CloudTrail only allow CMK to encypt with SSE, I foced use my CMK.
Let me check permission at my lab.

guitarrapc replied:
@tianchu Thanks, I've tested on my lab and confirm only `kms:Decrypt` is needed for downloading.
I've rebase and pushed fix.
can you review?

guitarrapc committed:
fix: remove none necessary permissiong

If you use a customer managed CMK, the Forwarder only needs kms:Decrypt for downloading, typically CloudTrail SSE require CMK and this permission will work for the case.
if you use the AWS managed CMK (i.e., key aws/s3), no extra permissions are required for the Forwarder to download the object.

tianchu reviewed (approved):
👍 LGTM!
